client/components/editor/teaching-elements/Assessment/TextSelection/Selections.js

```diff
@@ -0,0 +1,62 @@
+import { extractIndex } from './helpers';
+import cloneDeep from 'lodash/cloneDeep';
+import curry from 'lodash/curry';
+import curryRight from 'lodash/curryRight';
+import first from 'lodash/first';
+import inRange from 'lodash/inRange';
+import last from 'lodash/last';
+
+export default class Selections {
+  constructor(selection = []) {
+    this.ranges = cloneDeep(selection);
+  }
+
+  merge(selection) {
+    const begin = first(selection);
+    const final = last(selection);
+    const check = curry(isOverlaped);
+    const beginOverlap = this.ranges.find(check(begin));
+    const endOverlap = this.ranges.find(check(final - 1));
+    const range = [first(beginOverlap) || begin, last(endOverlap) || final];
+    this.remove(range);
+    this.ranges.push(range);
+    return range;
+  }
+
+  remove(range) {
+    const check = curryRight(isOverlaped)(range);
+    this.ranges = this.ranges.filter(range => (
+      !check(first(range)) &&
+      !check(last(range) - 1)
```

Comment:
Hmm, correct me if I'm wrong but this can be simplified by:
1. applying DeMorgan's law:
```js
range => !(check(first(range)) || check(last(range) - 1))
```
2. after which you can get rid of negation by _flipping_ operation itself
```js
reject(this.ranges, it => check(first(it)) || check(last(it) - 1);
```